### PR TITLE
Sync with direcrt upload

### PR DESCRIPTION
**Pull request by piotrze**
Fixes: https://github.com/mdyd-dev/pos-cli/issues/342

QA steps:

- [ ] sync asset file in
  - [ ] `app/assets/`
  - [ ] `app/assets/subdir/`
  - [ ] `app/assets/subdir/nextsubdir`
  - [ ] `modules/somemod/public/assets/`
  - [ ] `modules/somemod/public/assets/subdir/`
- [ ] take project with many asstets (ex. golfstix) and change all of them at once. You might cut assets files, start `sync` and then pase them in the same place.
- [ ] sync large asset file

### CODE REVIEW

File: lib/presignUrl.js [hunk 1]
```diff
@@ -4,9 +4,10 @@ const fs = require('fs'),
   mime = require('mime');
 
 const Portal = require('./portal');
+const deployServiceUrl = process.env.DEPLOY_SERVICE_URL ? process.env.DEPLOY_SERVICE_URL : `${process.env.MARKETPLACE_URL}/api/private/urls`;
```
Comment (pavelloz) [inline lib/presignUrl.js]:
```suggestion
const deployServiceUrl = process.env.DEPLOY_SERVICE_URL || `${process.env.MARKETPLACE_URL}/api/private/urls`;
```
?

File: lib/presignUrl.js [hunk 2]
```diff
@@ -29,4 +30,27 @@ const presignUrl = (s3FileName, fileName) => {
     });
 };
 
-module.exports = presignUrl;
+const presignDirectory = (path) => {
+  const serviceUrl = `${deployServiceUrl}/presign-directory`;
+  const params = { directory: path };
+
+  return request
```
Comment (pavelloz) [inline lib/presignUrl.js]: create method in proxy.js (or as its called across the system, gateway)

File: lib/watch.js [hunk 3]
```diff
@@ -56,20 +59,53 @@ const pushFile = (gateway, syncedFilePath) => {
...
+    logger.Debug(directUploadData);
+    await uploadFileFormData(filePath, directUploadData);
+    // const manifest = await generateManifest();
+    // logger.Debug(manifest);
+    // await gateway.sendManifest(manifest);
+    logger.Success(`[Sync] Synced asset: ${filePath}`);
+  } catch (e) {
+    logger.Debug(e.message);
+    logger.Debug(e.stack);
+    logger.Error('Sync assets failed.');
```
Comment (pavelloz) [inline lib/watch.js]:
```suggestion
    logger.Error(`[Sync] Failed to sync: ${filePath}`);
```

File: lib/watch.js [hunk 4]
```diff
@@ -56,20 +59,53 @@ const pushFile = (gateway, syncedFilePath) => {
...
+const pushFileDirectAssets = (gateway, syncedFilePath) => {
+  if (syncedFilePath.startsWith('app/assets') || moduleAssetRegex.test(syncedFilePath)){
+    sendFileToPresignedUrl(gateway, syncedFilePath)
+    return Promise.resolve(true);
+  } else {
+    return pushFile(gateway, syncedFilePath);
+  }
+};
+
+const sendFileToPresignedUrl = async (gateway, filePath) => {
```
Comment (pavelloz) [inline lib/watch.js]: looks like gateway is not necessary here

**Comment by Developer-DNM** [comment]
VERY exciting 👍

**Comment by jnorbert** [comment]
@piotrze I had a problem with the upload of assets in modules. It was a case where I have files with sizes of more than 50 MB.

Steps to reproduce:
1. In NM example copy to the assets files with size more than 50 MB
2. Make a deploy

In my case I've copy files to the /admincms/assets/documents.

**Comment by pavelloz** [comment]
Today i uploaded 10+ files ~20-90 MB using sync, so i guess you are not using feature from this PR.

**Comment by jnorbert** [comment]
@pavelloz Yes, but the deploy command should works as well, right?

**Comment by jnorbert** [comment]
@piotrze at the moment it's work only on staging right?
On production env, I got the error message

> pos-cli sync -d production                                                                                                                                                               
(node:82166) UnhandledPromiseRejectionWarning: StatusCodeError: 404 - "<!DOCTYPE html>\n<html lang=\"en\">\n<head>\n<meta charset=\"utf-8\">\n<title>Error</title>\n</head>\n<body>\n<pre>Cannot GET /presign-directory</pre>\n</body>\n</html>\n"
    at new StatusCodeError (/Users/jaceknorbert/pos-cli/node_modules/request-promise-core/lib/errors.js:32:15)
    at Request.plumbing.callback (/Users/jaceknorbert/pos-cli/node_modules/request-promise-core/lib/plumbing.js:104:33)
    at Request.RP$callback [as _callback] (/Users/jaceknorbert/pos-cli/node_modules/request-promise-core/lib/plumbing.js:46:31)
    at Request.self.callback (/Users/jaceknorbert/pos-cli/node_modules/request/request.js:185:22)
    at Request.emit (events.js:198:13)
    at Request.<anonymous> (/Users/jaceknorbert/pos-cli/node_modules/request/request.js:1161:10)
    at Request.emit (events.js:198:13)
    at IncomingMessage.<anonymous> (/Users/jaceknorbert/pos-cli/node_modules/request/request.js:1083:12)
    at Object.onceWrapper (events.js:286:20)
    at IncomingMessage.emit (events.js:203:15)
    at endReadableNT (_stream_readable.js:1145:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
(node:82166) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:82166) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**Pull request merged**